File: client/src/api/notebook.ts
import { apiFetch } from "./client";

export interface Project {
  id: string;
  code: string;
  name: string;
  color: string;
  status: string;
  owner_id: string | null;
}

export interface Experiment {
  id: string;
  project_id: string;
  code: string;
  title: string;
  status: string;
  locked_at: string | null;
  locked_by_id: string | null;
}

export interface EntryListItem {
  id: string;
  code: string;
  title: string;
  entry_date: string;
  tags: string[] | null;
  is_locked: boolean;
  experiment_id: string;
  created_at: string;
}

export interface EntryResult {
  id: string;
  entry_id: string;
  key: string;
  label: string;
  value_num: string | null;
  value_text: string | null;
  unit: string | null;
  tone: string;
}

export interface EntryAttachment {
  id: string;
  entry_id: string;
  kind: string;
  filename: string;
  original_name: string;
  mime_type: string | null;
  size_bytes: number | null;
  caption: string | null;
  storage_path: string;
  thumbnail_path: string | null;
}

export interface Entry {
  id: string;
  experiment_id: string;
  protocol_id: string | null;
  protocol_code: string | null;
  protocol_title: string | null;
  code: string;
  title: string;
  body_md: string | null;
  entry_date: string;
  tags: string[] | null;
  is_locked: boolean;
  created_by_id: string | null;
  created_at: string;
  updated_at: string;
  results: EntryResult[];
  attachments: EntryAttachment[];
}

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    const d = (err as { detail?: unknown }).detail;
    if (Array.isArray(d)) {
      throw new Error(d.map((e: { loc?: string[]; msg?: string }) => `${e.loc?.slice(-1)[0]}: ${e.msg}`).join(", "));
    }
    throw new Error(typeof d === "string" ? d : `Erreur ${res.status}`);
  }
  return res.json() as Promise<T>;
}

// Projects
export const listProjects = async (): Promise<Project[]> =>
  json(await apiFetch("/api/projects"));

export const createProject = async (data: { code: string; name: string; color?: string }): Promise<Project> =>
  json(await apiFetch("/api/projects", { method: "POST", body: JSON.stringify(data) }));

export const updateProject = async (id: string, data: Partial<Project>): Promise<Project> =>
  json(await apiFetch(`/api/projects/${id}`, { method: "PATCH", body: JSON.stringify(data) }));

export const deleteProject = async (id: string): Promise<void> => {
  const res = await apiFetch(`/api/projects/${id}`, { method: "DELETE" });
  if (!res.ok) throw new Error(`Erreur ${res.status}`);
};

// Experiments
export const listExperiments = async (projectId: string): Promise<Experiment[]> =>
  json(await apiFetch(`/api/projects/${projectId}/experiments`));

export const createExperiment = async (data: { project_id: string; code: string; title: string }): Promise<Experiment> =>
  json(await apiFetch("/api/experiments", { method: "POST", body: JSON.stringify(data) }));

export const getExperiment = async (id: string): Promise<Experiment> =>
  json(await apiFetch(`/api/experiments/${id}`));

export const deleteExperiment = async (id: string): Promise<void> => {
  const res = await apiFetch(`/api/experiments/${id}`, { method: "DELETE" });
  if (!res.ok) throw new Error(`Erreur ${res.status}`);
};

export const lockExperiment = async (id: string): Promise<Experiment> =>
  json(await apiFetch(`/api/experiments/${id}/lock`, { method: "POST" }));

// Entries
export const listEntries = async (expId: string): Promise<EntryListItem[]> =>
  json(await apiFetch(`/api/experiments/${expId}/entries`));

export const getEntry = async (id: string): Promise<Entry> =>
  json(await apiFetch(`/api/entries/${id}`));

export interface EntryHistory {
  id: string;
  entry_id: string;
  user_id: string | null;
  user_name: string | null;
  action: "created" | "updated" | "locked";
  changed_fields: Record<string, string> | null;
  created_at: string;
}

export const getEntryHistory = async (id: string): Promise<EntryHistory[]> =>
  json(await apiFetch(`/api/entries/${id}/history`));

export const createEntry = async (data: {
  experiment_id: string;
  code: string;
  title: string;
  body_md?: string | null;
  entry_date: string;
  tags?: string[];
  protocol_id?: string | null;
}): Promise<Entry> =>
  json(await apiFetch("/api/entries", { method: "POST", body: JSON.stringify(data) }));

export const updateEntry = async (id: string, data: {
  title?: string;
  body_md?: string;
  entry_date?: string;
  tags?: string[];
  protocol_id?: string | null;
}): Promise<Entry> =>
  json(await apiFetch(`/api/entries/${id}`, { method: "PATCH", body: JSON.stringify(data) }));

export const deleteEntry = async (id: string): Promise<void> => {
  const res = await apiFetch(`/api/entries/${id}`, { method: "DELETE" });
  if (!res.ok) throw new Error(`Erreur ${res.status}`);
};

export const lockEntry = async (id: string): Promise<Entry> =>
  json(await apiFetch(`/api/entries/${id}/lock`, { method: "POST" }));

// Results
export const addResult = async (entryId: string, data: {
  key: string; label: string; value_num?: number; value_text?: string; unit?: string; tone?: string;
}): Promise<EntryResult> =>
  json(await apiFetch(`/api/entries/${entryId}/results`, { method: "POST", body: JSON.stringify(data) }));

export const updateResult = async (entryId: string, resultId: string, data: {
  label?: string; value_num?: number | null; value_text?: string | null; unit?: string | null; tone?: string;
}): Promise<EntryResult> =>
  json(await apiFetch(`/api/entries/${entryId}/results/${resultId}`, { method: "PATCH", body: JSON.stringify(data) }));

export const deleteResult = async (entryId: string, resultId: string): Promise<void> => {
  const res = await apiFetch(`/api/entries/${entryId}/results/${resultId}`, { method: "DELETE" });
  if (!res.ok) throw new Error(`Erreur ${res.status}`);
};

File: server/app/schemas/notebook.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict


# ── Projects ───────────────────────────────────────────────────────────────────

class ProjectCreate(BaseModel):
    code: str
    name: str
    description: str | None = None
    color: str = "#1E5BC6"
    status: str = "active"


class ProjectUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    color: str | None = None
    status: str | None = None


class ProjectRead(BaseModel):
    id: UUID
    code: str
    name: str
    description: str | None
    color: str
    status: str
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class ProjectListItem(BaseModel):
    id: UUID
    code: str
    name: str
    color: str
    status: str
    owner_id: UUID | None
    model_config = ConfigDict(from_attributes=True)


# ── Experiments ────────────────────────────────────────────────────────────────

class ExperimentCreate(BaseModel):
    project_id: UUID
    code: str
    title: str
    description: str | None = None
    status: str = "running"


class ExperimentUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    status: str | None = None


class ExperimentRead(BaseModel):
    id: UUID
    project_id: UUID
    code: str
    title: str
    description: str | None
    status: str
    locked_at: datetime | None
    locked_by_id: UUID | None
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class ExperimentListItem(BaseModel):
    id: UUID
    project_id: UUID
    code: str
    title: str
    status: str
    locked_at: datetime | None
    locked_by_id: UUID | None
    model_config = ConfigDict(from_attributes=True)


# ── Entry Results ──────────────────────────────────────────────────────────────

class EntryResultCreate(BaseModel):
    key: str
    label: str
    value_num: Decimal | None = None
    value_text: str | None = None
    unit: str | None = None
    tone: str = "neutral"


class EntryResultUpdate(BaseModel):
    label: str | None = None
    value_num: Decimal | None = None
    value_text: str | None = None
    unit: str | None = None
    tone: str | None = None


class EntryResultRead(BaseModel):
    id: UUID
    entry_id: UUID
    key: str
    label: str
    value_num: Decimal | None
    value_text: str | None
    unit: str | None
    tone: str
    model_config = ConfigDict(from_attributes=True)


# ── Entry Attachments ──────────────────────────────────────────────────────────

class EntryAttachmentRead(BaseModel):
    id: UUID
    entry_id: UUID
    kind: str
    filename: str
    original_name: str
    mime_type: str | None
    size_bytes: int | None
    caption: str | None
    storage_path: str
    thumbnail_path: str | None
    model_config = ConfigDict(from_attributes=True)


# ── Notebook Entries ───────────────────────────────────────────────────────────

class NotebookEntryCreate(BaseModel):
    experiment_id: UUID
    code: str
    title: str
    body_md: str | None = None
    entry_date: date
    tags: list[str] | None = None
    protocol_id: UUID | None = None


class NotebookEntryUpdate(BaseModel):
    title: str | None = None
    body_md: str | None = None
    entry_date: date | None = None
    tags: list[str] | None = None
    protocol_id: UUID | None = None


class NotebookEntryListItem(BaseModel):
    id: UUID
    code: str
    title: str
    entry_date: date
    tags: list[str] | None
    is_locked: bool
    experiment_id: UUID
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class NotebookEntryRead(BaseModel):
    id: UUID
    experiment_id: UUID
    protocol_id: UUID | None
    protocol_code: str | None = None
    protocol_title: str | None = None
    code: str
    title: str
    body_md: str | None
    entry_date: date
    tags: list[str] | None
    is_locked: bool
    created_by_id: UUID | None
    created_at: datetime
    updated_at: datetime
    results: list[EntryResultRead] = []
    attachments: list[EntryAttachmentRead] = []
    model_config = ConfigDict(from_attributes=True)


class EntryHistoryRead(BaseModel):
    id: UUID
    entry_id: UUID
    user_id: UUID | None
    user_name: str | None
    action: str
    changed_fields: dict | None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)
